tests/integration/tables/beta/test_tpp.py

```diff
@@ -0,0 +1,236 @@
+    in_memory_engine.populate(
+        # Simple case: successive registrations
+        {
+            tpp.practice_registrations: [
+                dict(
+                    patient_id=1,
+                    practice_pseudo_id=123,
+                    start_date=date(2000, 1, 1),
+                    end_date=date(2005, 1, 1),
+                ),
+                dict(
+                    patient_id=1,
+                    practice_pseudo_id=456,
+                    start_date=date(2005, 1, 1),
+                    end_date=date(2015, 1, 1),
+                ),
+                dict(
+                    patient_id=1,
+                    practice_pseudo_id=789,
+                    start_date=date(2015, 1, 1),
+                    end_date=date(2020, 1, 1),
+                ),
+            ]
+        },
+        # Registration with NULL end date
+        {
+            tpp.practice_registrations: [
+                dict(
+                    patient_id=2,
+                    practice_pseudo_id=123,
+                    start_date=date(2000, 1, 1),
+                    end_date=None,
+                ),
+            ]
+        },
+        # Overlapping: choose most recent
+        {
+            tpp.practice_registrations: [
+                dict(
+                    patient_id=3,
+                    practice_pseudo_id=456,
+                    start_date=date(2000, 1, 1),
+                    end_date=date(2015, 1, 1),
+                ),
+                dict(
+                    patient_id=3,
+                    practice_pseudo_id=789,
+                    start_date=date(2005, 1, 1),
+                    end_date=date(2015, 1, 1),
+                ),
+            ]
+        },
+        # Overlapping: choose longest
+        {
+            tpp.practice_registrations: [
+                dict(
+                    patient_id=4,
+                    practice_pseudo_id=123,
+                    start_date=date(2000, 1, 1),
+                    end_date=date(2012, 1, 1),
+                ),
+                dict(
+                    patient_id=4,
+                    practice_pseudo_id=456,
+                    start_date=date(2000, 1, 1),
+                    end_date=date(2015, 1, 1),
+                ),
+            ]
+        },
+        # Tie-break: choose largest pseudo ID
+        {
+            tpp.practice_registrations: [
+                dict(
+                    patient_id=5,
+                    practice_pseudo_id=789,
+                    start_date=date(2000, 1, 1),
+                    end_date=date(2015, 1, 1),
+                ),
+                dict(
+                    patient_id=5,
+                    practice_pseudo_id=456,
+                    start_date=date(2000, 1, 1),
+                    end_date=date(2015, 1, 1),
+                ),
+            ]
+        },
+    )
+
+    active_reg = tpp.practice_registrations.active_for_patient_at("2010-01-01")
```

Comment:
Just following from the comment above to help me think it through, if the test patient would have: `end_date=date(2010, 1, 1),` the current function `active_for_patient_at ()` would not include that patient, and I _think_ that's exactly what we want?

Reply:
To be honest, I don't actually know! My assumption was that a registration period includes the start date and all the days up to _but not including_ the end date. But quite possibly it includes the end date as well. We might be able to tell by looking at the data.

Reply:
I have changed this to include the end date. If the end date overlaps with the start date of the next registration this makes no difference as we'll prefer the newer registration in any case.